Use pgx 0.6.0 with PostgreSQL 15 support

Smittyvb opened the pull request.

epgts commented on .github/workflows/ci.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
     strategy:
       matrix:
-        pgversion: [12, 13, 14]
+        pgversion: [12, 13, 14, 15]
```
We need to break this up into separate merges.
1. Push updated image (#639 also needs to go in, and #621 is harmless and nice to have).
2. This line and all the rest.

epgts replied:
Never mind; it's not worth the bother to put both 0.5 and 0.6 support into the image, so I'm happy to just push an updated image from this branch before merging it.

epgts commented on docker/ci/setup.sh:
```diff
@@ -127,7 +127,7 @@ EOF
                     postgresql$pg-devel \
                     postgresql$pg-server \
-                    timescaledb-2-postgresql-$pg
+                    # timescaledb-2-postgresql-$pg
```
Oof.  A catch-22.

If we can't wait for timescaledb to release pg15 support first, we can make this install this for all but 15 and also skip doctester for 15.  If we go that route, file an issue and leave a TODO comment referencing the issue.

Smittyvb replied:
I updated the CI to skip running doctests/update tests (turns out both of those have a dependency on TimescaleDB). Issue: #648.

epgts commented on tools/dependencies.sh:
```diff
@@ -7,14 +7,14 @@
+PGX_VERSION=0.6.0
 
-RUST_TOOLCHAIN=1.60.0
+RUST_TOOLCHAIN=1.65.0
```
Dang, really?  Do they really require 1.65 now? :(

epgts replied:
Hmm, for some reason they removed rust-version from their Cargo.toml files so it's back to getting confusing and potentially misleading error messages when building with too old a compiler?

Smittyvb replied:
The `rust-version` in the Cargo.toml files were never accurate anyways. My testing has revealed that it fails on 1.60 and works on 1.65, so I just went the most recent version – I think the actual MSRV might be 1.63 or 1.61 though.

epgts replied:
pgx 0.6.0 builds fine here on 1.60.0 .

We are starting to fall behind, for sure.  I suggest we update to 1.64.0 in 1.14, not in this release.

Smittyvb replied:
It [failed in CI on 1.60](https://github.com/timescale/timescaledb-toolkit/actions/runs/3593608416/jobs/6050778282):

```
   Compiling pgx v0.6.0
error[E0658]: use of unstable library feature 'slice_ptr_len'
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/pgx-0.6.0/src/slice.rs:28:32
   |
28 |         index.le(&self.pallocd.len()).then(|| self.get_unchecked(index))
   |                                ^^^
   |
   = note: see issue #71146 <https://github.com/rust-lang/rust/issues/71146> for more information
```

and that feature was [partially stabilized in 1.63](https://github.com/rust-lang/rust/pull/94640) – I think 1.63 is the MSRV. I'll do some testing.

epgts replied:
> The rust-version in the Cargo.toml files were never accurate anyways

They were accurate when they were born:  https://github.com/tcdi/pgx/pull/575

It's a shame to leave dependents to reverse-engineer the requirements by trying various versions.  Maybe we can ask them to bring them back and add verifying them to the release checklist?

Also, I think I tested the wrong thing:  I tested only building cargo-pgx, not pgx.  Whoops!

Updating to 1.64 or 1.63 here is fine by me.

Thanks!

Smittyvb replied:
I updated the Rust version used to 1.63.

Smittyvb commented:
The CI [now passes on this branch](https://github.com/timescale/timescaledb-toolkit/actions/runs/3597095618/jobs/6058498106), although we don't run doc/upgrade tests for pg15 yet.

epgts commented on tools/dependencies.sh:
```diff
@@ -7,14 +7,15 @@
 
-PG_VERSIONS='12 13 14'
+PG_VERSIONS='12 13 14 15'
+TSDB_PG_VERSIONS='12 13 14'
```
TODO on fixing this too?

epgts reviewed (commented):
Looks good!

Smittyvb commented:
Updated the PR to pgx 0.6.1, released 2 hours ago. [The CI still passes](https://github.com/timescale/timescaledb-toolkit/actions/runs/3632917873) when run with the Docker image from this branch.

Smittyvb commented:
I rebased and updated the changelog entry to indicate that PostgreSQL 15 support is experimental. I ran the CI again, [it still passes](https://github.com/timescale/timescaledb-toolkit/actions/runs/3679211848).

epgts reviewed (changes requested):
Actually, hang on, what's wrong with Clippy?

rtwalker commented:
> Actually, hang on, what's wrong with Clippy?

I think the new pgx is using a feature that was stabilized in Rust 1.63 which our older (1.60?) clippy thinks is unstable.

Smittyvb commented:
The error with Clippy happens when you try to compile pgx with rustc < 1.63, since rustc 1.63 stabilized a library feature that pgx relies on:
```
error[E0658]: use of unstable library feature 'slice_ptr_len'
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/pgx-0.6.1/src/slice.rs:28:32
   |
28 |         index.le(&self.pallocd.len()).then(|| self.get_unchecked(index))
   |                                ^^^
   |
   = note: see issue #71146 <https://github.com/rust-lang/rust/issues/71146> for more information
```

This PR updates the Docker image to 1.63 which does support the needed feature.

epgts reviewed (approved):
Of course.  When Brian approves this, you can publish the image from this branch and then the checks will clear?  Thanks!

Smittyvb commented:
bors r+
